classes/plain.lua

```diff
@@ -260,20 +248,110 @@ function class:registerCommands ()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.parindent", SILE.nodefactory.glue())
+      SILE.settings:set("current.parindent", SILE.nodefactory.glue())
+      SILE.settings:set("document.lskip", SILE.nodefactory.hfillglue(lskip.width.length))
+      SILE.settings:set("document.rskip", SILE.nodefactory.hfillglue(rskip.width.length))
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
+      SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0))
+      SILE.process(content)
+      SILE.call("par")
+    end)
+  end, "Typeset its contents in a centered block (keeping margins).")
+
   self:registerCommand("raggedright", function (_, content)
-    SILE.call("ragged", { right = true }, content)
-  end)
+    SILE.settings:temporarily(function ()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.lskip", SILE.nodefactory.glue(lskip.width.length))
+      SILE.settings:set("document.rskip", SILE.nodefactory.hfillglue(rskip.width.length))
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
+      SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0))
+      SILE.process(content)
+      SILE.call("par")
+    end)
+  end, "Typeset its contents in a left aligned block (keeping margins).")
 
   self:registerCommand("raggedleft", function (_, content)
-    SILE.call("ragged", { left = true }, content)
+    SILE.settings:temporarily(function ()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.lskip", SILE.nodefactory.hfillglue(lskip.width.length))
+      SILE.settings:set("document.rskip", SILE.nodefactory.glue(rskip.width.length))
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
+      SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0))
+      SILE.process(content)
+      SILE.call("par")
+    end)
+  end, "Typeset its contents in a right aligned block (keeping margins).")
+
+  self:registerCommand("justified", function (_, content)
+    SILE.settings:temporarily(function ()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.lskip", SILE.nodefactory.glue(lskip.width.length))
+      SILE.settings:set("document.rskip", SILE.nodefactory.glue(rskip.width.length))
+      SILE.settings:set("document.spaceskip", nil)
+      -- HACK. This knows too much about parfillskip defaults...
+      -- (Which must be big, but smaller than infinity. Doh!)
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue("0pt plus 10000pt"))
```

Comment:
I overlooked the possibility to reset a setting to its default value, which should do the trick here.

Reply:
Done.